### PR TITLE
Soc 7590 line endings

### PR DESCRIPTION
**Pull request by PeterGerrard**
Handle line endings in a way such that if given `\r\r\n` it is consistently parsed so a hunk header exception is not thrown.

We do this by not canonicolising line endings in the patch and handling \r\n in the parsing of individual components.

We have a bunch of tests already around line endings but this issue wasn't being caught as it is abount diff producing 1 patch and it being parsed as a slightly different patch. We fix this by reverting a change in #6 and instead we do manual checks in each parse rather than assuming we have read with pure unix line endings.

### CODE REVIEW

File: NGit.Test/NGit.Patch/PatchTest.cs [hunk 1]
```diff
@@ -80,15 +80,15 @@ public virtual void TestParse_ConfigCaseInsensitive()
...
 			FileHeader fRepositoryConfigTest = p.GetFiles()[0];
 			FileHeader fRepositoryConfig = p.GetFiles()[1];
 			NUnit.Framework.Assert.AreEqual("org.eclipse.jgit.test/tst/org/spearce/jgit/lib/RepositoryConfigTest.java"
 				, fRepositoryConfigTest.GetNewPath());
 			NUnit.Framework.Assert.AreEqual("org.eclipse.jgit/src/org/spearce/jgit/lib/RepositoryConfig.java"
 				, fRepositoryConfig.GetNewPath());
-			NUnit.Framework.Assert.AreEqual(572, fRepositoryConfigTest.startOffset);
-			NUnit.Framework.Assert.AreEqual(1490, fRepositoryConfig.startOffset);
+			NUnit.Framework.Assert.AreEqual(m_Crlf ? 586 : 572, fRepositoryConfigTest.startOffset);
+			NUnit.Framework.Assert.AreEqual(m_Crlf ? 1520 : 1490, fRepositoryConfig.startOffset);
```
Comment (angellaa) [inline NGit.Test/NGit.Patch/PatchTest.cs:91]: What are all this magic numbers? Can we extract them in constants?

**Comment by angellaa** [comment]
There are various places where we look behind in buffers... is there any chance we can get overflow exceptions? should we be a bit more defensive?

File: NGit.Test/NGit.Api/ApplyCommandTest.cs [hunk 2]
```diff
@@ -204,5 +204,29 @@ public virtual void TestModifyZ()
...
+            File.WriteAllText(modifiedFile, post);
+            git.Diff().SetOutputStream(stream).Call();
+
+            stream.Position = 0;
+            File.WriteAllText(modifiedFile, pre);
+
+            git.Apply().SetPatch(stream).Call();
+
+            Assert.That(File.ReadAllText(modifiedFile), Is.EqualTo(post));
+        }
```
Comment (angellaa) [inline NGit.Test/NGit.Api/ApplyCommandTest.cs]: This test is not following the pattern that all other tests in the suite seems to follow (having a preImage, postImage and a patch in the NGit.Diff and later do assertions on the ApplyResult). This test runs 4 times but m_UseCrlfPatches and m_UseCrlfFiles are ignored because we do not call Init. Any reason why we can't make this test following the same patters of the others? If we can't (how can we create a patch file?) maybe it belongs to a separate test suite (we don't want to run it 4 times).

File: NGit/NGit.Api/ApplyCommand.cs [hunk 3]
```diff
@@ -227,9 +226,11 @@ private void Apply(FilePath f, FileHeader fh)
 			}
 			IList<string> newLines = new AList<string>(oldLines);
 		    RawText hrt = null;
+		    var containedCRLF = false;
             foreach (HunkHeader hh in fh.GetHunks())
 			{
 				var buffer = Sharpen.Runtime.GetStringForBytes(hh.GetBuffer(), hh.GetStartOffset(), hh.GetEndOffset() - hh.GetStartOffset());
+			    containedCRLF |= buffer.Contains("\r\n");
```
Comment (tomcsmith1990) [inline NGit/NGit.Api/ApplyCommand.cs]: I'm not sure why the changes here are needed?

File: NGit.Test/NGit.Patch/PatchTest.cs (hunk 1 above)
Comment (tomcsmith1990) [inline NGit.Test/NGit.Patch/PatchTest.cs:91]: I'm not sure why we need the if statement in here now either? Isn't the point that the patches get applied in the same way regardless of the line endings, which means these shouldn't change?

File: NGit.Test/NGit.Api/ApplyCommandTest.cs (hunk 2 above)
Comment (angellaa) [inline NGit.Test/NGit.Api/ApplyCommandTest.cs]: I had a play with it and tried to conform the test with the others but it is apparent that this is a special thing and useCrlfFiles does not make sense for this test because it is explicitly testing an edge case. When we do, we get a different error. Also having the actual files in the solution makes the test less intuitive (it's not obvious what the content of the file is and also Visual Studio prompt to fix the line endings that it is not useful in this case). So after talking with Matt we decided to stick with the original test. I am going to move this test out of this class.

File: NGit/NGit.Patch/BinaryHunk.cs [hunk 4]
```diff
@@ -168,7 +168,7 @@ internal virtual int ParseHunk(int ptr, int end)
 			//
 			while (ptr < end)
 			{
-				bool empty = buf[ptr] == '\n';
+				bool empty = (buf[ptr] == '\n') || (buf[ptr] == '\r' && buf[ptr-1] == '\n');
```
Comment (angellaa) [inline NGit/NGit.Patch/BinaryHunk.cs]: The test does not cover this change, so appears to be unnecessary. Why are we doing it? If it is important, can we please add a test for this? also I suspect that this might throws an exception if there is only one character in the buffer.

File: NGit/NGit.Patch/FileHeader.cs [hunk 5]
```diff
@@ -711,6 +715,10 @@ private static string P1(string r)
 		internal virtual FileMode ParseFileMode(int ptr, int end)
 		{
 			int tmp = 0;
+		    if (buf[end - 2] == '\r') // Ignore windows line ending
```
Comment (angellaa) [inline NGit/NGit.Patch/FileHeader.cs]: I suspect this is a way to detect that the line ending is \r\r\n. I would probably extract this in a method and test all the three characters (with overflow checks) and also reuse it in the other places where you do buf[dot2 - 2] == '\r'

**Comment by angellaa** [comment]
Overall I am not very confident about these changes and it is difficult to review as I didn't have done the work and it is unclear what the purpose of every change is (some changes seems not strictly related to the new test added). Should we talk about the changes face to face?

File: NGit.Test/NGit.Patch/PatchTest.cs (hunk 1 above)
Comment (PeterGerrard) [inline NGit.Test/NGit.Patch/PatchTest.cs:91]: We did have it assuming that all Windows Line Endings would be converted into Unix Line endings. We now don't do this as it causes issues with `\r\r\n` and instead we just use the original line endings.

So in this test the location of each `\n` will vary depending on if there are windows line endings or not.

File: NGit/NGit.Api/ApplyCommand.cs (hunk 3 above)
Comment (PeterGerrard) [inline NGit/NGit.Api/ApplyCommand.cs]: We need to know if there if there are any windows line endings in the patch. so instead of doing it when parsing we do it when we read in from the buffer.

File: NGit/NGit.Patch/BinaryHunk.cs (hunk 4 above)
Comment (PeterGerrard) [inline NGit/NGit.Patch/BinaryHunk.cs]: This is tested by an existing test. The change we made to `Patch.Parse` (reverting an old change) causes this parsing to fail. So we fix this to ensure the tests written for the original issue still pass.

File: NGit/NGit.Patch/FileHeader.cs (hunk 5 above)
Comment (PeterGerrard) [inline NGit/NGit.Patch/FileHeader.cs]: Its not a check for `\r\r\n` but I'll go have a shot at pulling out a well named method.

**Comment by PeterGerrard** [comment]
I'm happy to have a chat, and I'll try and some more info into the first comment.

**Comment by tomcsmith1990** [comment]
I do wonder what kind of performance impact this has since we're now looking at lots of strings more, but migrations is already slow enough that I'd prefer to take a performance hit right now to get the correct behaviour #brokenwindowsyndrome

**Comment by angellaa** [comment]
Thank for the chat Peter, I now have a bit better understanding of the changes. As we agreed, please add a quick summary of the decision we made in the pull request. I am happy to merge the changes after you did that. The performance consideration is interesting.

**Comment by PeterGerrard** [comment]
I don't think we the performance will take too much of a hit as the canonaclisation still had to do all that parsing for \r\n

**Comment by angellaa** [comment]
I can do a quick performance testing if you want....

**Comment by PeterGerrard** [comment]
@angellaa Added the summary.

**Comment by angellaa** [comment]
Just for reference I measured the time required to apply the patch for a patch of 10000 lines of code (around 16 megabytes) and here are the results. Now it takes about 0.0441635 seconds while previously it took 0.0000574 seconds. So the solution is three order of magnitude slower but still pretty quick. I am going to commit an optimization that makes the new implementation as quick as before for patch files with only windows line endings (that is the majority of our customers).
